### DISCUSSION TITLE
Can't get (playground-tested) GraphQL mutation working in SDK

### DISCUSSION THREAD

**Discussion by fredgrieco** [rubrikinc/rubrik-powershell-sdk#45] [Q&A] [answered]
I have the following mutation working in the API playground that triggers an on demand backup.  I have not been able to get this to work in the Powershell SDK though (best attempt and errors, bottom).  I am pretty sure it's getting hung up on the var syntax.
How should I be doing this?  Is there an example of a mutation with the "invoke-rsc" command anywhere?

working:
mutation TakeAzureVMSnapshotMutation (\$input: StartCreateAzureNativeVirtualMachineSnapshotsJobInput!) {
 startCreateAzureNativeVirtualMachineSnapshotsJob(input: \$input) {
   jobIds {jobId    __typename }
    errors { error  __typename  }  __typename  }}
"@{"input": { "virtualMachineRubrikIds":[ "11111111-1111-1111-1111-1111111111111" ], "retentionSlaId": "22222222-2222-2222-2222-22222222222"}}

Powershell -- not working:

 Invoke-Rsc -GqlQuery "mutation TakeAzureVMSnapshotMutation ($input : StartCreateAzureNativeVirtualMachineSnapshotsJobInput!) {startCreateAzureNativeVirtualMachineSnapshotsJob(input : $input) {jobIds {jobId __typename} errors {error __typename} __typename}" -Var '{"input"  :  {
  "virtualMachineRubrikIds" : ["11111111-1111-1111-1111-1111111111111"],"retentionSlaId" : "22222222-2222-2222-2222-22222222222"}}'

Invoke-Rsc : Cannot bind parameter 'Var'. Cannot convert the "{"input"  :  {"virtualMachineRubrikIds" : ["11111111-1111-1111-1111-1111111111111" ],
  "retentionSlaId" : "22222222-2222-2222-2222-22222222222"}}" value of type "System.String" to type "System.Collections.Hashtable".

At line:1 char:263
+ ... typename} errors {error __typename} __typename}" -Var '{"input"  :  {
+                                                           ~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Invoke-Rsc], ParameterBindingException
    + FullyQualifiedErrorId : CannotConvertArgumentNoMessage,RubrikSecurityCloud.PowerShell.Cmdlets.Invoke_Rsc

**Comment by doslagerRubrik** [comment]
Try putting your mutation in a var or @ here string. It looks like there are GraphQL variables in your mutation so they may be getting evaluated as powershell variables. Put the mutation into a var, write-host the variable to make sure it's what you want it to be. The do the invoke with the var. 

Also, you have double quotes inside a double quote, so the query as you are presenting it is getting chopped. I changed the quotes around the var to single quote instead of double quote since there are already double quote inside the string which is JSON.

EDIT: I updated the query to single quote for the string as well. Now it's working. If you put single quotes, the string will be whatever is inside the pair of quotes, if you put double quotes and there is text that looks like a powershell var (ie $input), it will try to evaluate what $input should be, but since it's $null, it just outputs nothing instead of the string $input.




try something like:

```
$query = 'mutation TakeAzureVMSnapshotMutation ($input : StartCreateAzureNativeVirtualMachineSnapshotsJobInput!) {startCreateAzureNativeVirtualMachineSnapshotsJob(input : $input) {jobIds {jobId __typename} errors {error __typename} __typename}'


$var = '{"input" : {
"virtualMachineRubrikIds" : ["11111111-1111-1111-1111-1111111111111"],"retentionSlaId" : "22222222-2222-2222-2222-22222222222"}}'

write-host $query
write-host $var
<confirm $query and $var are outputting to what you want>


invoke-rsc -GqlQuery $query -var $var
```


Also, to confirm the $var is correct, you can convertto-json and confirm it converts properly to a hashtable 

```
 $var | ConvertFrom-Json -Depth 100

input
-----
@{virtualMachineRubrikIds=System.Object[]; retentionSlaId=22222222-2222-2222-2222-22222222222}
```

**Comment by clumnah** [comment, marked as answer]
Your account team will be reaching out with this same info soon. 

**Make sure you are on the most up to date version of the SDK, as there have been lots of changes.** 

When I first looked at your query, your approach made logical sense. I ran it internally and got an error telling me you could not take an on demand snapshot for that workload. I then looked at the API code capture content from an on-demand snapshot via the UI and noticed it was using a different query. When I asked why, it is because not all objects have been implemented into the generic TakeOnDemand snapshot query operation yet. So...the better approach is to use the object specific query instead. 

So the change I made was to go from 
`$query = New-RscMutationSnapshot -Operation TakeOnDemand` 
to 
`$query = New-RscMutationVsphere -Operation OnDemandSnapshot`

I also prefer to use the objects just from an ease of use perspective. Its easier to read and troubleshoot. 

```
Import-Module RubrikSecurityCloud
#region Variables
$VMId = "a8650f5c-c5d6-5fdb-9245-e288668fd296"
$SLADomain = "1d-14d-NRA"
#endregion
Connect-Rsc

$RscVsphereVm = Get-RscVmwareVm -Id $VMId
$RscSlaDomain = Get-RscSla -Name $SLADomain 
$query = New-RscMutation -GqlMutation vsphereOnDemandSnapshot
$query.var.input = New-Object -TypeName RubrikSecurityCloud.Types.VsphereOnDemandSnapshotInput
$query.var.input.config = New-Object -TypeName RubrikSecurityCloud.Types.BaseOnDemandSnapshotConfigInput
$query.var.input.config.slaId = $RscSlaDomain.id
$query.var.input.id = $RscVsphereVm.Id
$query.invoke()
```

Hope this helps.

**Reply by fredgrieco** [reply]
Thanks Chris!  Yes, account team did reach out to me about this.  Some details and caveats for my case:

-  Make sure you have the latest module release as this is "pre 1.0" and changes are being made regularly.  For instance, "Get-RscSlaDomain" was replaced by "Get-RscSla"
- Use Powershell 7.  PS 5 would not take the object build above and failed on some .net-looking error.

- For AzureVM, the query (mutation) build was as follows:

```
$mutation = New-RscMutation -GqlMutation startCreateAzureNativeVirtualMachineSnapshotsJob -Addfield jobIds.jobid
$mutation.var.input = New-Object -TypeName RubrikSecurityCloud.Types.startCreateAzureNativeVirtualMachineSnapshotsJob Input
$mutation.var.input.config = New-Object -TypeName RubrikSecurityCloud.Types.BaseOnDemandSnapshotConfigInput
$mutation.var.input.config.retentionslaId = $RscSlaDomain.id
$mutation.var.input.VirtualMachineRubrikIds= $RscVsphereVm.Id   /// or "@($vmID")
Invoke-Rsc -Query $mutation
```
